File: src/main/java/org/example/kalkulationsprogramm/service/WebPushService.java
package org.example.kalkulationsprogramm.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import nl.martijndwars.webpush.Notification;
import nl.martijndwars.webpush.PushService;
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.example.kalkulationsprogramm.domain.KalenderEintrag;
import org.example.kalkulationsprogramm.domain.Mitarbeiter;
import org.example.kalkulationsprogramm.domain.PushSubscription;
import org.example.kalkulationsprogramm.repository.KalenderEintragRepository;
import org.example.kalkulationsprogramm.repository.MitarbeiterRepository;
import org.example.kalkulationsprogramm.repository.PushSubscriptionRepository;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import jakarta.annotation.PostConstruct;
import java.security.GeneralSecurityException;
import java.security.Security;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;
import java.util.*;

@Slf4j
@Service
@RequiredArgsConstructor
public class WebPushService {

    private final PushSubscriptionRepository pushSubscriptionRepository;
    private final KalenderEintragRepository kalenderEintragRepository;
    private final MitarbeiterRepository mitarbeiterRepository;
    private final ObjectMapper objectMapper;

    @Value("${push.vapid.public-key:}")
    private String vapidPublicKey;

    @Value("${push.vapid.private-key:}")
    private String vapidPrivateKey;

    @Value("${push.vapid.subject:mailto:noreply@handwerk-erp.de}")
    private String vapidSubject;

    @Value("${zeiterfassung.base-url:https://localhost}")
    private String baseUrl;

    private PushService pushService;

    // Track sent notifications to avoid duplicates (key: "appointmentId_type")
    private final Set<String> sentNotifications = Collections.synchronizedSet(new HashSet<>());
    private static final DateTimeFormatter DATE_FORMATTER = DateTimeFormatter.ofPattern("EEE, dd. MMM", Locale.GERMAN);

    @PostConstruct
    public void init() {
        if (vapidPublicKey.isBlank() || vapidPrivateKey.isBlank()) {
            log.warn("Web Push VAPID keys not configured. Push notifications disabled. " +
                    "Generate keys and set PUSH_VAPID_PUBLIC_KEY / PUSH_VAPID_PRIVATE_KEY environment variables.");
            return;
        }

        try {
            Security.addProvider(new BouncyCastleProvider());
            pushService = new PushService(vapidPublicKey, vapidPrivateKey, vapidSubject);
            log.info("Web Push Service initialized with VAPID keys");
        } catch (GeneralSecurityException e) {
            log.error("Failed to initialize Web Push Service: {}", e.getMessage());
        }
    }

    public boolean isEnabled() {
        return pushService != null;
    }

    public String getVapidPublicKey() {
        return vapidPublicKey;
    }

    /**
     * Subscribe a device for push notifications.
     */
    @Transactional
    public void subscribe(Long mitarbeiterId, String endpoint, String p256dh, String auth) {
        // Remove existing subscription for this endpoint (re-subscribe)
        pushSubscriptionRepository.findByEndpoint(endpoint).ifPresent(existing -> {
            pushSubscriptionRepository.delete(existing);
            pushSubscriptionRepository.flush();
        });

        Mitarbeiter mitarbeiter = mitarbeiterRepository.findById(mitarbeiterId)
                .orElseThrow(() -> new IllegalArgumentException("Mitarbeiter not found: " + mitarbeiterId));

        PushSubscription sub = new PushSubscription();
        sub.setMitarbeiter(mitarbeiter);
        sub.setEndpoint(endpoint);
        sub.setP256dh(p256dh);
        sub.setAuth(auth);
        pushSubscriptionRepository.save(sub);

        log.info("Push subscription saved for Mitarbeiter {}", mitarbeiterId);
    }

    /**
     * Unsubscribe a device.
     */
    @Transactional
    public void unsubscribe(String endpoint) {
        pushSubscriptionRepository.deleteByEndpoint(endpoint);
        log.info("Push subscription removed for endpoint");
    }

    /**
     * Scheduled task: Check every 2 minutes for upcoming appointments
     * and send push notifications 24h and 1h before.
     */
    @Scheduled(fixedDelay = 120_000, initialDelay = 30_000)
    public void checkAndSendNotifications() {
        if (!isEnabled()) return;

        try {
            LocalDate today = LocalDate.now();
            LocalDate tomorrow = today.plusDays(1);
            LocalDate dayAfter = today.plusDays(2);

            // Fetch appointments for today, tomorrow, and day after (covers 24h window)
            List<KalenderEintrag> appointments = kalenderEintragRepository.findByDatumBetween(today, dayAfter);

            LocalDateTime now = LocalDateTime.now();

            for (KalenderEintrag apt : appointments) {
                LocalDateTime aptDateTime = getAppointmentDateTime(apt);
                if (aptDateTime.isBefore(now)) continue;

                long minutesUntil = ChronoUnit.MINUTES.between(now, aptDateTime);

                // 24h notification: between 23h50m and 24h10m before
                if (minutesUntil >= 1430 && minutesUntil <= 1450) {
                    sendNotificationForAppointment(apt, "24h");
                }

                // 1h notification: between 50min and 70min before
                if (minutesUntil >= 50 && minutesUntil <= 70) {
                    sendNotificationForAppointment(apt, "1h");
                }
            }

            // Cleanup old sent-tracking entries (older than 3 days)
            cleanupSentTracking();
        } catch (Exception e) {
            log.error("Error checking appointment notifications: {}", e.getMessage());
        }
    }

    private LocalDateTime getAppointmentDateTime(KalenderEintrag apt) {
        if (apt.isGanztaegig() || apt.getStartZeit() == null) {
            return apt.getDatum().atTime(8, 0); // Default 08:00 for all-day events
        }
        return apt.getDatum().atTime(apt.getStartZeit());
    }

    private void sendNotificationForAppointment(KalenderEintrag apt, String type) {
        String trackingKey = apt.getId() + "_" + type + "_" + apt.getDatum();
        if (sentNotifications.contains(trackingKey)) return;

        // Find all mitarbeiter who should receive this notification
        Set<Long> mitarbeiterIds = new HashSet<>();

        // Ersteller (creator)
        if (apt.getErsteller() != null) {
            mitarbeiterIds.add(apt.getErsteller().getId());
        }

        // Teilnehmer (participants)
        if (apt.getTeilnehmer() != null) {
            for (Mitarbeiter t : apt.getTeilnehmer()) {
                mitarbeiterIds.add(t.getId());
            }
        }

        // If no specific people assigned (company calendar), send to all active employees
        if (mitarbeiterIds.isEmpty()) {
            mitarbeiterRepository.findAll().stream()
                    .filter(m -> Boolean.TRUE.equals(m.getAktiv()))
                    .forEach(m -> mitarbeiterIds.add(m.getId()));
        }

        // Build notification payload
        String title;
        String body;
        String timeStr = apt.isGanztaegig() || apt.getStartZeit() == null
                ? "Ganztägig"
                : apt.getStartZeit().format(DateTimeFormatter.ofPattern("HH:mm")) + " Uhr";

        if ("24h".equals(type)) {
            title = "Termin morgen: " + apt.getTitel();
            body = apt.getDatum().format(DATE_FORMATTER) + " um " + timeStr;
        } else {
            title = "Termin in 1 Stunde: " + apt.getTitel();
            body = timeStr;
        }

        String notificationUrl = baseUrl + "/zeiterfassung/kalender?termin=" + apt.getId();

        // Send to all subscriptions for relevant mitarbeiter
        for (Long mitarbeiterId : mitarbeiterIds) {
            List<PushSubscription> subscriptions = pushSubscriptionRepository.findByMitarbeiterId(mitarbeiterId);
            for (PushSubscription sub : subscriptions) {
                sendPush(sub, title, body, notificationUrl, apt.getId(), type);
            }
        }

        sentNotifications.add(trackingKey);
    }

    private void sendPush(PushSubscription sub, String title, String body, String url, Long appointmentId, String type) {
        try {
            Map<String, Object> payload = new HashMap<>();
            payload.put("title", title);
            payload.put("body", body);
            payload.put("url", url);
            payload.put("appointmentId", appointmentId);
            payload.put("type", type);
            payload.put("timestamp", System.currentTimeMillis());

            String payloadJson = objectMapper.writeValueAsString(payload);

            Notification notification = new Notification(
                    sub.getEndpoint(),
                    sub.getP256dh(),
                    sub.getAuth(),
                    payloadJson.getBytes()
            );

            pushService.send(notification);
            log.debug("Push sent to endpoint for appointment {}: {}", appointmentId, title);
        } catch (Exception e) {
            String errorMsg = e.getMessage() != null ? e.getMessage() : "";
            // 410 Gone or 404: subscription expired, remove it
            if (errorMsg.contains("410") || errorMsg.contains("404") || errorMsg.contains("expired")) {
                log.info("Removing expired push subscription: {}", sub.getEndpoint().substring(0, Math.min(50, sub.getEndpoint().length())));
                try {
                    pushSubscriptionRepository.delete(sub);
                } catch (Exception deleteErr) {
                    log.warn("Failed to delete expired subscription: {}", deleteErr.getMessage());
                }
            } else {
                log.warn("Failed to send push notification: {}", errorMsg);
            }
        }
    }

    private void cleanupSentTracking() {
        // Keep tracking set manageable - remove entries for past dates
        LocalDate yesterday = LocalDate.now().minusDays(1);
        sentNotifications.removeIf(key -> {
            try {
                String datePart = key.substring(key.lastIndexOf('_') + 1);
                LocalDate entryDate = LocalDate.parse(datePart);
                return entryDate.isBefore(yesterday);
            } catch (Exception e) {
                return true; // remove malformed entries
            }
        });
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/service/WebPushService.kt
package org.example.kalkulationsprogramm.service

import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.annotation.PostConstruct
import nl.martijndwars.webpush.Notification
import nl.martijndwars.webpush.PushService
import org.bouncycastle.jce.provider.BouncyCastleProvider
import org.example.kalkulationsprogramm.domain.KalenderEintrag
import org.example.kalkulationsprogramm.domain.Mitarbeiter
import org.example.kalkulationsprogramm.domain.PushSubscription
import org.example.kalkulationsprogramm.repository.KalenderEintragRepository
import org.example.kalkulationsprogramm.repository.MitarbeiterRepository
import org.example.kalkulationsprogramm.repository.PushSubscriptionRepository
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.security.KeyFactory
import java.security.KeyPair
import java.security.Security
import java.security.interfaces.ECPublicKey
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.X509EncodedKeySpec
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import java.util.Base64
import java.util.Collections
import java.util.Locale

@Service
class WebPushService(
    private val pushSubscriptionRepository: PushSubscriptionRepository,
    private val kalenderEintragRepository: KalenderEintragRepository,
    private val mitarbeiterRepository: MitarbeiterRepository,
    private val objectMapper: ObjectMapper,
) {
    @Value("\${push.vapid.public-key:}")
    private lateinit var configuredVapidPublicKey: String

    @Value("\${push.vapid.private-key:}")
    private lateinit var vapidPrivateKey: String

    @Value("\${push.vapid.subject:mailto:noreply@handwerk-erp.de}")
    private lateinit var vapidSubject: String

    @Value("\${zeiterfassung.base-url:https://localhost}")
    private lateinit var baseUrl: String

    private var pushService: PushService? = null
    private var rawVapidPublicKey: String? = null
    private val sentNotifications: MutableSet<String> = Collections.synchronizedSet(HashSet())

    val isEnabled: Boolean
        get() = pushService != null

    val vapidPublicKey: String
        get() = rawVapidPublicKey.orEmpty()

    @PostConstruct
    fun init() {
        if (configuredVapidPublicKey.isBlank() || vapidPrivateKey.isBlank()) {
            log.warn(
                "Web Push VAPID keys not configured. Push notifications disabled. " +
                    "Generate keys and set PUSH_VAPID_PUBLIC_KEY / PUSH_VAPID_PRIVATE_KEY environment variables.",
            )
            return
        }

        try {
            Security.addProvider(BouncyCastleProvider())
            val pubBytes = Base64.getUrlDecoder().decode(configuredVapidPublicKey)
            val privBytes = Base64.getUrlDecoder().decode(vapidPrivateKey)

            val keyFactory = KeyFactory.getInstance("EC", "BC")
            val publicKey = keyFactory.generatePublic(X509EncodedKeySpec(pubBytes))
            val privateKey = keyFactory.generatePrivate(PKCS8EncodedKeySpec(privBytes))
            val keyPair = KeyPair(publicKey, privateKey)

            val ecPub = publicKey as ECPublicKey
            val x = ecPub.w.affineX.toByteArray()
            val y = ecPub.w.affineY.toByteArray()
            val rawPoint = ByteArray(65)
            rawPoint[0] = 0x04
            System.arraycopy(x, maxOf(0, x.size - 32), rawPoint, 1 + maxOf(0, 32 - x.size), minOf(32, x.size))
            System.arraycopy(y, maxOf(0, y.size - 32), rawPoint, 33 + maxOf(0, 32 - y.size), minOf(32, y.size))
            rawVapidPublicKey = Base64.getUrlEncoder().withoutPadding().encodeToString(rawPoint)

            pushService = PushService(keyPair, vapidSubject)
            log.info("Web Push Service initialized with VAPID keys")
        } catch (e: Exception) {
            log.error("Failed to initialize Web Push Service: {}", e.message, e)
        }
    }

    fun notifyAll(title: String, body: String, url: String) {
        if (!isEnabled) {
            log.debug("WebPush nicht aktiv – notifyAll wird ignoriert")
            return
        }
        try {
            val alle = pushSubscriptionRepository.findAll()
            for (sub in alle) {
                sendPush(sub, title, body, url, null, "freigabe")
            }
        } catch (e: Exception) {
            log.warn("notifyAll fehlgeschlagen: {}", e.message)
        }
    }

    fun notifyFreigabeAnnahme(title: String, body: String, url: String) {
        if (!isEnabled) {
            log.debug("WebPush nicht aktiv – notifyFreigabeAnnahme wird ignoriert")
            return
        }
        try {
            val alle = pushSubscriptionRepository.findAll()
            for (sub in alle) {
                val ma = sub.mitarbeiter ?: continue
                val erlaubt = ma.abteilungen?.any { it.darfFreigabeAnnahmePushen == true } == true
                if (!erlaubt) continue
                sendPush(sub, title, body, url, null, "freigabe")
            }
        } catch (e: Exception) {
            log.warn("notifyFreigabeAnnahme fehlgeschlagen: {}", e.message)
        }
    }

    fun notifyWebseitenAnfrage(title: String, body: String, url: String) {
        if (!isEnabled) {
            log.debug("WebPush nicht aktiv – notifyWebseitenAnfrage wird ignoriert")
            return
        }
        try {
            val alle = pushSubscriptionRepository.findAll()
            for (sub in alle) {
                val ma = sub.mitarbeiter ?: continue
                val erlaubt = ma.abteilungen?.any { it.darfWebseitenAnfragenPushen == true } == true
                if (!erlaubt) continue
                sendPush(sub, title, body, url, null, "anfrage")
            }
        } catch (e: Exception) {
            log.warn("notifyWebseitenAnfrage fehlgeschlagen: {}", e.message)
        }
    }

    @Transactional
    fun subscribe(mitarbeiterId: Long?, endpoint: String?, p256dh: String?, auth: String?) {
        requireNotNull(mitarbeiterId) { "Mitarbeiter ID fehlt" }
        require(!endpoint.isNullOrBlank()) { "Endpoint fehlt" }
        require(!p256dh.isNullOrBlank()) { "p256dh fehlt" }
        require(!auth.isNullOrBlank()) { "auth fehlt" }

        pushSubscriptionRepository.findByEndpoint(endpoint).ifPresent { existing ->
            pushSubscriptionRepository.delete(existing)
            pushSubscriptionRepository.flush()
        }

        val mitarbeiter = mitarbeiterRepository.findById(mitarbeiterId)
            .orElseThrow { IllegalArgumentException("Mitarbeiter not found: $mitarbeiterId") }

        val sub = PushSubscription().apply {
            this.mitarbeiter = mitarbeiter
            this.endpoint = endpoint
            this.p256dh = p256dh
            this.auth = auth
        }
        pushSubscriptionRepository.save(sub)
        log.info("Push subscription saved for Mitarbeiter {}", mitarbeiterId)
    }

    @Transactional
    fun unsubscribe(endpoint: String) {
        pushSubscriptionRepository.deleteByEndpoint(endpoint)
        log.info("Push subscription removed for endpoint")
    }

    @Scheduled(fixedDelay = 120_000, initialDelay = 30_000)
    fun checkAndSendNotifications() {
        if (!isEnabled) return

        try {
            val today = LocalDate.now()
            val dayAfter = today.plusDays(2)
            val appointments = kalenderEintragRepository.findByDatumBetween(today, dayAfter)
            val now = LocalDateTime.now()

            for (apt in appointments) {
                val aptDateTime = getAppointmentDateTime(apt)
                if (aptDateTime.isBefore(now)) continue

                val minutesUntil = ChronoUnit.MINUTES.between(now, aptDateTime)
                if (minutesUntil in 1430..1450) {
                    sendNotificationForAppointment(apt, "24h")
                }
                if (minutesUntil in 50..70) {
                    sendNotificationForAppointment(apt, "1h")
                }
            }

            cleanupSentTracking()
        } catch (e: Exception) {
            log.error("Error checking appointment notifications: {}", e.message)
        }
    }

    private fun getAppointmentDateTime(apt: KalenderEintrag): LocalDateTime =
        if (apt.isGanztaegig() || apt.startZeit == null) {
            (apt.datum ?: LocalDate.now()).atTime(8, 0)
        } else {
            (apt.datum ?: LocalDate.now()).atTime(apt.startZeit)
        }

    private fun sendNotificationForAppointment(apt: KalenderEintrag, type: String) {
        val trackingKey = "${apt.id}_${type}_${apt.datum}"
        if (sentNotifications.contains(trackingKey)) return

        val mitarbeiterIds = HashSet<Long>()
        apt.ersteller?.id?.let { mitarbeiterIds.add(it) }
        apt.teilnehmer?.forEach { teilnehmer -> teilnehmer.id?.let { mitarbeiterIds.add(it) } }

        if (mitarbeiterIds.isEmpty()) {
            mitarbeiterRepository.findAll()
                .filter { it.aktiv == true }
                .forEach { it.id?.let(mitarbeiterIds::add) }
        }

        val timeStr =
            if (apt.isGanztaegig() || apt.startZeit == null) {
                "Ganztägig"
            } else {
                val startZeit = apt.startZeit
                "${startZeit?.format(DateTimeFormatter.ofPattern("HH:mm"))} Uhr"
            }

        val title: String
        val body: String
        if (type == "24h") {
            title = "Termin morgen: ${apt.titel}"
            body = "${(apt.datum ?: LocalDate.now()).format(DATE_FORMATTER)} um $timeStr"
        } else {
            title = "Termin in 1 Stunde: ${apt.titel}"
            body = timeStr
        }

        val notificationUrl = "$baseUrl/zeiterfassung/kalender?termin=${apt.id}"
        for (mitarbeiterId in mitarbeiterIds) {
            val subscriptions = pushSubscriptionRepository.findByMitarbeiterId(mitarbeiterId)
            for (sub in subscriptions) {
                sendPush(sub, title, body, notificationUrl, apt.id, type)
            }
        }

        sentNotifications.add(trackingKey)
    }

    private fun sendPush(
        sub: PushSubscription,
        title: String,
        body: String,
        url: String,
        appointmentId: Long?,
        type: String,
    ) {
        try {
            val payload = HashMap<String, Any?>()
            payload["title"] = title
            payload["body"] = body
            payload["url"] = url
            payload["appointmentId"] = appointmentId
            payload["type"] = type
            payload["timestamp"] = System.currentTimeMillis()

            val payloadJson = objectMapper.writeValueAsString(payload)
            val notification = Notification(
                sub.endpoint,
                sub.p256dh,
                sub.auth,
                payloadJson.toByteArray(),
            )

            pushService?.send(notification)
            log.debug("Push sent to endpoint for appointment {}: {}", appointmentId, title)
        } catch (e: Exception) {
            val errorMsg = e.message ?: ""
            if (errorMsg.contains("410") || errorMsg.contains("404") || errorMsg.contains("expired")) {
                val endpoint = sub.endpoint.orEmpty()
                log.info("Removing expired push subscription: {}", endpoint.substring(0, minOf(50, endpoint.length)))
                try {
                    pushSubscriptionRepository.delete(sub)
                } catch (deleteErr: Exception) {
                    log.warn("Failed to delete expired subscription: {}", deleteErr.message)
                }
            } else {
                log.warn("Failed to send push notification: {}", errorMsg)
            }
        }
    }

    private fun cleanupSentTracking() {
        val yesterday = LocalDate.now().minusDays(1)
        sentNotifications.removeIf { key ->
            try {
                val datePart = key.substring(key.lastIndexOf('_') + 1)
                val entryDate = LocalDate.parse(datePart)
                entryDate.isBefore(yesterday)
            } catch (e: Exception) {
                true
            }
        }
    }

    private companion object {
        private val log = LoggerFactory.getLogger(WebPushService::class.java)
        private val DATE_FORMATTER: DateTimeFormatter = DateTimeFormatter.ofPattern("EEE, dd. MMM", Locale.GERMAN)
    }
}
